ambition_utils/anomaly/models.py

```diff
@@ -0,0 +1,257 @@
+import datetime
+import sys
+
+
+from django.db import models
+from django.contrib.postgres.fields import JSONField
+from django.utils.functional import cached_property
+import fleming
+from tdigest import TDigest
+
+from manager_utils import ManagerUtilsManager, ManagerUtilsQuerySet
+
+
+class BadPercentileValue(Exception):
+    pass
+
+
+class BadAnomalyType(Exception):
+    pass
+
+
+class AnomalyBaseQueryset(ManagerUtilsQuerySet):
+    def unprocessed(self, assume_now=None):
+        if assume_now is None:
+            assume_now = datetime.datetime.utcnow()
+        today = fleming.floor(assume_now, day=1)
+
+        return self.filter(last_modified__lt=today)
+
+
+class AnomalyBaseManager(ManagerUtilsManager):
+    def get_queryset(self):
+        return AnomalyBaseQueryset(self.model)
+
+    def unprocessed(self, assume_now=None):
+        return self.get_queryset().unprocessed(assume_now=assume_now)
+
+
+class AnomalyBase(models.Model):
+    """
+    Override this class to create an Anomaly Detector.  Think of this as a key-value store
+    for anomalies.  You update instances with observed data.  This teaches the detector what
+    "normal" is.  You then check values against what it has learned using the .check method.
+
+    You are free to add whatever fields you want to your detector as long as they don't
+    collide with the field names on this base class.
+
+    You must define a get_uid() method on your class to populate the uid field.
+    """
+    # set whether or not this detector will use the blob to do incremental updates
+    # When inheriting from this class, you must specify whether or not this detector will be incremental.
+    # Incremental detectors use the tdigest blob.  Non-incremental detectors rely on outside code for updates
+    # (for example, direct sql computation)
+    IS_INCREMENTAL = True
+    PERCENTILE_LOW_DEFAULT = 5
+    PERCENTILE_HIGH_DEFAULT = 95
+
+    class Meta:
+        abstract = True
+
+    # unique id to identify this anomaly detector
+    uid = models.CharField(max_length=256, unique=True)
```

Comment:
Could this be a uuid field? or are we treating this more like a name?

Reply:
We are treating this like a name

Reply:
We are treating it like a name.  I can change it to something like `name` or `unique_name`, but I thought a `str` uid field would properly describe it's purpose.

Reply:
Yeah its fine to keep it